compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4086,6 +4097,20 @@ public void visit(BLangBinaryExpr binaryExpr) {
         resultType = types.checkType(binaryExpr, actualType, expType);
     }
 
+    private boolean isNullableBinaryExpr(BLangBinaryExpr binaryExpr) {
+        if (binaryExpr.expectedType.isNullable() && binaryExpr.expectedType.tag != TypeTags.ANY) {
```

Comment:
How/why is the expected type relevant here?

Reply:
This function is used on L4021 to check whether the expected type is `float?` or `decimal?`